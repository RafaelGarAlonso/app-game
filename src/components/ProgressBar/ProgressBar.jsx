import { useEffect, useState } from 'react';
import { ProgressBar } from 'react-bootstrap';

export const AnimatedProgressBar = ({ onComplete, timeToComplete = 25, variant }) => {
    const [loadingProgress, setLoadingProgress] = useState(0);

    useEffect(() => {
        const counter = setInterval(() => {
            setLoadingProgress((c) => {
                const newValue = c + 1;
                if (newValue <= 100) return newValue;
                if (newValue > 100) {
                    clearInterval(counter);
                    onComplete();
                }
            });
        }, timeToComplete);
        return () => {
            clearInterval(counter);
            setLoadingProgress(0);
        }
    }, [onComplete, timeToComplete]);

    return (
        <>
            <ProgressBar variant={variant} animated now={loadingProgress} />
        </>
    )
}