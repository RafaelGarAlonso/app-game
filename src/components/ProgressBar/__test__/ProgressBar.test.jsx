import { render, waitFor } from '@testing-library/react';
import { AnimatedProgressBar } from '../ProgressBar';
import { act } from 'react';

describe('Test <ProgressBar />', () => {

    const onComplete = jest.fn();

    beforeEach(() => {
        jest.useFakeTimers();
        jest.spyOn(global, 'setInterval');
        jest.spyOn(global, 'clearInterval');
    });

    test('Check snapshot component', () => {
        const { container } = render(<AnimatedProgressBar />);
        expect(container).toMatchSnapshot();
        ;
    });

    test('Check ProgressBar fills in time', async () => {
        const { container } = render(<AnimatedProgressBar />);
        const delay = 1000;

        act(() => {
            jest.advanceTimersByTime(delay - 500);
        });

        await waitFor(() => {
            const progressBar = container.querySelector(".progress-bar");
            expect(progressBar.getAttribute("style")).toBe("width: 58%;");
        });
    });

    test('Check ProgressBar when finished time', async () => {
        render(<AnimatedProgressBar onComplete={onComplete} />);
        const delay = 3000;

        act(() => {
            jest.advanceTimersByTime(delay);
        });

        await waitFor(() => {
            expect(onComplete).toHaveBeenCalled();
        });
    });

});