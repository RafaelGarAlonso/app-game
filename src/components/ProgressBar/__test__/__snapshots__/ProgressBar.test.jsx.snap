// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ProgressBar /> Check snapshot component 1`] = `
<div>
  <div
    class="progress"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="progress-bar progress-bar-animated progress-bar-striped"
      role="progressbar"
      style="width: 0%;"
    />
  </div>
</div>
`;
